### PR TITLE
Add custom regex with groups support

### LINKED ISSUE #98: [path] Custom regex with numbered matches
**Is this a BUG REPORT or FEATURE REQUEST?**:
feature

**Anything else we need to know?**
regex with numbered groups extracting 2 query strings
`re=\\?query=(?P<2>[^&]+)\\&more=(?P<1>[^&]+)`: _redirect.string.more.example.com
Can't get the above to split it into 2 groups in Go. Have it working on https://regexr.com/ so.

?P<1> is used to name the group, which we could use for order.

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Adds custom regex with groups support

**Which issue this PR fixes**:
fixes #98

### CODE REVIEW

File: path_test.go [hunk 1]
```diff
@@ -125,6 +125,14 @@ func Test_zoneFromPath(t *testing.T) {
 			"_redirect.string.example.com",
 			nil,
 		},
+		{
+			"example.com",
+			"/12345-some-path?query=string&more=stuff",
+			"",
+			"\\?query=(?P<a>[^&]+)\\&more=(?P<b>[^&]+)",
```
Comment (stp-ip) [inline path_test.go]: We should do a more complicated test that actually reorders the matches.

``` suggestion
			"\\?query=(?P<b>[^&]+)\\&more=(?P<a>[^&]+)\\&test=(?P<c>[^&]+)",
```

Additionally we should test for the correct handling of some errors such as using the same ordered match multiple times.

Comment (erbesharat) [reply path_test.go]: Will add more tests and error handlings.

File: path_test.go [hunk 2]
```diff
@@ -125,6 +125,30 @@ func Test_zoneFromPath(t *testing.T) {
...
+			"_redirect.string.stuff.example.com",
+			nil,
+		},
+		{
+			"example.com",
+			"/12345-some-path?query=string&more=stuff&test=testing",
+			"",
+			"\\?query=(?P<b>[^&]+)\\&more=(?P<a>[^&]+)\\&test=(?P<c>[^&]+)",
+			"_redirect.stuff.string.testing.example.com",
```
Comment (stp-ip) [inline path_test.go]: Shouldn't the ordering be:
```
"_redirect.testing.string.stuff"
```
So the query matches `string` as `b`, `stuff` as `a` and `testing` as `c`.
Using our usual ordering would mean _redirect.c.b.a.example.com.
Am I missing something?

Comment (erbesharat) [reply path_test.go]: Forgot to use ``reverse()``. Changed the tests and used reverse in the last commit.

**Comment by stp-ip** [comment]
Merging with my own review only.

**Pull request merged**